File: oplog.js
/*
console.log(db.getSiblingDB("local").getCollection("oplog.rs").aggregate([
  {
    $match: { ts: {$gte: Timestamp( Math.floor((new Date().getTime() / 1000) - 5), 0)}},
  },
  { $match: {
    $or: [
        { op: { $in: ["i", "u", "d"] }},
        {
            op: "c",
            ns: "admin.$cmd",
            "o.applyOps": {$type: "array"},
        },
    ],
  } },
  {
    $addFields: {
      ops: {
        $cond: {
          if: { $eq: ["$op", "c"] },
          then: "$o.applyOps",
          else: ["$$ROOT"]
        }
      }
    }
  },
  { $unwind: "$ops" },
  {
    $match: {
      "ops.op": { $in: ["i", "u", "d"] }
    }
  },
  {
    $addFields: {
      size: { $bsonSize: "$ops" }
    }
  },
  {
    $group: {
      _id: "$ops.op",
      count: { $sum: 1 },
      totalSize: { $sum: "$size" }
    }
  },
  {
    $project: {
      op: "$_id",
      count: 1,
      totalSize: 1,
      _id: 0
    }
  }
]).toArray())
*/

function opAgg(opRef) {
    return {
        $cond: {
            if: {$eq: [opRef + ".op", "u"]},
            then: {
                $concat: [
                    opRef + ".op",
                    "/",
                    {$cond: {
                        if: {$eq: ["missing", {$type: opRef + ".o._id"}]},
                        then: "u",
                        else: "r",
                    }},
                ],
            },
            else: opRef + ".op",
        },
    }
}

let cursor = db.getSiblingDB("local").getCollection("oplog.rs").find(
  {
    $or: [
        {
            op: { $in: ["i", "u", "d"] },
        },
        {
            op: "c",
            ns: "admin.$cmd",
            "o.applyOps": {$type: "array"},
        },
    ],
  },
  {
    op: opAgg("$$ROOT"),
    size: {
        $cond: {
            if: { $eq: ["$op", "c"] },
            then: "$$REMOVE",
            else: { $bsonSize: "$$ROOT" },
        },
    },
    ops: {
        $cond: {
            if: { $eq: ["$op", "c"] },
            then: {
                $map: {
                    input: "$o.applyOps",
                    as: "opEntry",
                    in: {
                        size: {$bsonSize: "$$opEntry"},
                        op: opAgg("$$opEntry"),
                    },
                },
            },
            else: "$$REMOVE",
        },
    },
    ts: 1,
  },
);

cursor.tailable({awaitData: true});

while (true) {
    console.log(cursor.next());
}

File: oplog.go
package main

import (
	"github.com/FGasper/mongo-measure-writes/agg"
	"github.com/samber/lo"
	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
)

var eventsToTruncate = sliceOf(
	"insert",
	"update",
	"replace",
	"delete",
)

var oplogOps = sliceOf("i", "u", "d")

var oplogQuery = bson.D{
	{"$or", bson.A{
		bson.D{
			{"op", bson.D{{"$in", oplogOps}}},
		},
		bson.D{
			{"op", "c"},
			{"ns", "admin.$cmd"},
			{"o.applyOps", bson.D{{"$type", "array"}}},
		},
	}},
}

func makeOpFieldExpr(opRef string) bson.D {
	return agg.Cond{
		If: agg.In(opRef+".op", "u", "applyOps.u"),
		Then: agg.Concat(
			opRef+".op",
			"/",
			agg.Cond{
				If:   agg.Eq("missing", agg.Type(opRef+".o._id")),
				Then: "u",
				Else: "r",
			},
		),
		Else: opRef + ".op",
	}.D()
}

var oplogUnrollFormatStages = mongo.Pipeline{
	// Match op in ["i", "u", "d"] or applyOps array
	{{"$match", oplogQuery}},

	// Normalize to ops array
	{{"$addFields", bson.D{
		{"ops", bson.D{
			{"$cond", bson.D{
				{"if", bson.D{{"$eq", bson.A{"$op", "c"}}}},
				{"then", bson.D{
					{"$map", bson.D{
						{"input", "$o.applyOps"},
						{"as", "opEntry"},
						{"in", bson.D{
							{"$mergeObjects", bson.A{
								"$$opEntry",
								bson.D{
									{"op", bson.D{
										{"$concat", bson.A{
											"applyOps.",
											"$$opEntry.op",
										}},
									}},
								},
							}},
						}},
					}},
				}},
				{"else", bson.A{"$$ROOT"}},
			}},
		}},
	}}},

	// Unwind ops
	{{"$unwind", "$ops"}},

	// Match sub-ops of interest
	{{"$match", bson.D{
		{"ops.op", bson.D{{"$in", lo.Flatten(
			sliceOf(
				oplogOps,
				lo.Map(
					oplogOps,
					func(op string, _ int) string { return "applyOps." + op },
				),
			),
		),
		}}},
	}}},

	// Add BSON size per op, and append either /u or /r to op:u.
	{{"$addFields", bson.D{
		{"size", bson.D{{"$bsonSize", "$ops"}}},
		{"ops.op", makeOpFieldExpr("$ops")},
	}}},
}
